python/gigl/experimental/knowledge_graph_embedding/lib/model/heterogeneous_graph_model.py

```diff
@@ -0,0 +1,439 @@
+        """
+        Infer node embeddings for a given NodeBatch.
+
+        Args:
+            node_batch (NodeBatch): The batch of nodes to infer embeddings for.
+
+        Returns:
+            torch.Tensor: The inferred node embeddings.
+        """
+        # Fetch node embeddings from the embedding layer
+        num_nodes = node_batch.nodes.values().numel()
+        node_embeddings_kt: torchrec.KeyedTensor = self.large_embeddings(
+            node_batch.nodes
+        )
+        node_embeddings = node_embeddings_kt.values()
+
+        node_embeddings = node_embeddings.reshape(
+            num_nodes, -1, self.node_emb_dim
+        )  # [num_nodes, num_node_types, node_dim]
+        node_embeddings = node_embeddings.sum(dim=1)  # [num_nodes, node_dim]
+
+        operator = (
+            self.src_operator
+            if self.phase == ModelPhase.INFERENCE_SRC
+            else self.dst_operator
+        )
+
+        # Apply the operator to the node embeddings
+        node_embeddings = operator(
+            embeddings=node_embeddings,
+            condensed_edge_types=node_batch.condensed_edge_type.repeat(num_nodes),
+        )
+        return node_embeddings
+
+    def forward(
+        self, edge_batch: EdgeBatch
+    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        # Fetch node embeddings from the embedding layer
+        src_embeddings, dst_embeddings = self.fetch_src_and_dst_embeddings(edge_batch)
+        condensed_edge_types = edge_batch.condensed_edge_types
+        labels = edge_batch.labels
+
+        # Apply relation operator to transform embeddings based on edge types
+        src_embeddings, dst_embeddings = self.apply_relation_operator(
+            src_embeddings=src_embeddings,
+            dst_embeddings=dst_embeddings,
+            condensed_edge_types=condensed_edge_types,
+        )
+
+        pos_src_embeddings, batch_neg_src_embeddings = (
+            src_embeddings[: self.active_sampling_config.positive_edge_batch_size],
+            src_embeddings[self.active_sampling_config.positive_edge_batch_size :],
+        )
+        pos_dst_embeddings, batch_neg_dst_embeddings = (
+            dst_embeddings[: self.active_sampling_config.positive_edge_batch_size],
+            dst_embeddings[self.active_sampling_config.positive_edge_batch_size :],
+        )
+        pos_condensed_edge_types, batch_neg_condensed_edge_types = (
+            condensed_edge_types[
+                : self.active_sampling_config.positive_edge_batch_size
+            ],
+            condensed_edge_types[
+                self.active_sampling_config.positive_edge_batch_size :
+            ],
+        )
+        pos_labels, batch_neg_labels = (
+            labels[: self.active_sampling_config.positive_edge_batch_size],
+            labels[self.active_sampling_config.positive_edge_batch_size :],
+        )
+
+        pos_logits: torch.Tensor
+        pos_labels: torch.Tensor
+        neg_logits = list()
+        neg_labels = list()
```

Comment:
nit, can we add types here? 
```suggestion
        neg_logits: list[torch.Tensor] = list()
        neg_labels: list[torch.Tensor] = list()
```